### ISSUE TITLE
Lifetime bounds are wrongly marked as erroneous in initial positions for trait objects

### ISSUE THREAD

**Issue by albel727** [intellij-rust/intellij-rust#2225]
All of the following is valid in Rust, but presently the variants with lifetime in front are highlighted like invalid, red-underlining the lifetime in question.

```rust
type WorksOk<'a> = Box<AsRef<str> + 'a>;
type ShowsAsError<'a> = Box<'a + AsRef<str>>;

type WorksOkAgain<'a> = Box<Fn() -> () + 'a>;
type ShowsAsErrorAgain<'a> = Box<'a + Fn() -> ()>;
```

It appears that there is conflict with normal parameter parsing, since errors are like

`Wrong number of lifetime parameters: expected 0, found 1 [E0107]`
`Wrong number of type parameters: expected 1, found 0 [E0243]`
